### PR TITLE
initial functions

### PR DESCRIPTION
**Pull request by puneethreddy20**

### CODE REVIEW

File: .idea/vcs.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (cviecco) [inline .idea/vcs.xml]: This whole file is unneeded. Please remove

File: main.go [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+
+type RuntimeState struct {
+	Config       AppConfigFile
+	//htmlTemplate *template.Template
+}
+
+const ldapTimeoutSecs = 3
+
+
+var Attributes = []string{"sAMAccountName"}
```
Comment (cviecco) [inline main.go]: The name of this variable is too Generic. Attributes for what?

File: main.go [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+	TLSKeyFilename        string `yaml:"tls_key_filename"`
+	StorageURL            string `yaml:"storage_url"`
+	OpenIDCConfigFilename string `yaml:"openidc_config_filename"`
+}
+
+type UserInfoLDAPSource struct {
+	BindUsername       string   `yaml:"bind_username"`
+	BindPassword       string   `yaml:"bind_password"`
+	LDAPTargetURLs     string   `yaml:"ldap_target_urls"`
+	UserSearchBaseDNs  string `yaml:"user_search_base_dns"`
```
Comment (cviecco) [inline main.go]: Please pass your sources over go fmt before passing for review.

File: main.go [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+	//htmlTemplate *template.Template
+}
+
+const ldapTimeoutSecs = 3
+
+
+var Attributes = []string{"sAMAccountName"}
+
+
+const pagingsize =2147483647
```
Comment (cviecco) [inline main.go]: Ditto with this constant.. make the name more specific.

File: main.go [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+
+
+func getLDAPConnection(u url.URL, timeoutSecs uint, rootCAs *x509.CertPool) (*ldap.Conn, string, error) {
+
+	if u.Scheme != "ldaps" {
+		err := errors.New("Invalid ldap scheme (we only support ldaps)")
+		return nil, "", err
+	}
+
+	//hostnamePort := server + ":636"
```
Comment (cviecco) [inline main.go]: remove unused code

File: main.go [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+		log.Printf("Source=%s", source)
+		return state, err
+	}
+	return state,err
+}
+
+
+
+
+func getLDAPConnection(u url.URL, timeoutSecs uint, rootCAs *x509.CertPool) (*ldap.Conn, string, error) {
```
Comment (cviecco) [inline main.go]: you took my function and used it as a model. However you dont use rootCA in your function. So either use it or remove it from the function signature

File: main.go [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+
+//establishing the connection.
+func getLDAPConnection1(u url.URL) (*ldap.Conn, string, error) {
+
+	if u.Scheme != "ldaps" {
+		err := errors.New("Invalid ldap scheme (we only support ldaps)")
+		return nil, "", err
+	}
+
+	//hostnamePort := server + ":636"
```
Comment (cviecco) [inline main.go]: remove dead code

File: main.go [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+	port := "636"
+
+	if len(serverPort) == 2 {
+		port = serverPort[1]
+	}
+
+	server := serverPort[0]
+	hostnamePort := server + ":" + port
+
+	//timeout := time.Duration(time.Duration(timeoutSecs) * time.Second)
```
Comment (cviecco) [inline main.go]: no timeout? ... there is a reason I added timeouts everywhere. All connections can take unbounded time to finish. You do not want this. All network activity must have timeouts

File: main.go [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+
+	// we dont close the tls connection directly  close defer to the new ldap connection
+	//conn := ldap.NewConn(tlsConn, true)
+	return conn, server, nil
+}
+
+
+
+//Function which returns the array of disabled accounts from Active Directory AD.
+func disabledaccounts(conn *ldap.Conn, UserSearchBaseDNs string, UserSearchFilter string, Attributes []string)([]string,error){
```
Comment (cviecco) [inline main.go]: The name of this function does not match us function. Please rename the function and use camelCase for function names.

File: main.go [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+	var final []string
```
Comment (cviecco) [inline main.go]: too generic function name

File: main.go [hunk 11]
```diff
@@ -0,0 +1,463 @@
...
+//function which compares the users disabled accounts in AD with CPE LDAP and adds the attribute nsaccountLock in CPE ldap for the disbaled USer.
+func disableinCPELDAP(conn *ldap.Conn,userdn string,result []string)(error){
+	for _,entry:=range result{
+		entry=userDN(userdn,entry)
+
+		modify := ldap.NewModifyRequest(entry)
+		modify.Replace("nsaccountLock",[]string{"True"})
+		err:=conn.Modify(modify)
+		if err !=nil{
+			return err
```
Comment (cviecco) [inline main.go]: before giving up, try to add the attribute(maybe I cannot add, becase the attribute is not there in the first place)

File: main.go [hunk 12]
```diff
@@ -0,0 +1,463 @@
...
+
+
+
+//function which compares the users disabled accounts in AD with CPE LDAP and adds the attribute nsaccountLock in CPE ldap for the disbaled USer.
+func disableinCPELDAP(conn *ldap.Conn,userdn string,result []string)(error){
+	for _,entry:=range result{
+		entry=userDN(userdn,entry)
+
+		modify := ldap.NewModifyRequest(entry)
+		modify.Replace("nsaccountLock",[]string{"True"})
```
Comment (cviecco) [inline main.go]: "nsaccountLock" should be a constant

File: main.go [hunk 13]
```diff
@@ -0,0 +1,463 @@
...
+	//fmt.Printf("%+v",result.Entries)
+	//fmt.Println(result.Entries[0].GetAttributeValue("uid"))
+	for _,entry := range result.Entries{
+		//if entry.GetAttributeValue("uid")
+		uid := entry.GetAttributeValue("uid")
+		final[uid]=uid
+
+
+	}
+	fmt.Println(final)
```
Comment (cviecco) [inline main.go]: this debug print should be in a debug block.

File: main.go [hunk 14]
```diff
@@ -0,0 +1,463 @@
...
+	searchrequest := ldap.NewSearchRequest(UserSearchBaseDNs,ldap.ScopeWholeSubtree,ldap.NeverDerefAliases, 0,0,false,UserSearchFilter,hello,nil)
+	result, err := conn.Search(searchrequest)
+	if err !=nil{
+		return nil,err
+	}
+	fmt.Println(len(result.Entries))
+	if len(result.Entries)==0{
+		log.Println("No records found")
+	}
+	//fmt.Printf("%+v",result.Entries)
```
Comment (cviecco) [inline main.go]: remove dead code

File: main.go [hunk 15]
```diff
@@ -0,0 +1,463 @@
...
+	}
+	return nil
+
+}
+
+
+
+
+//Get all ldap users and put that in map
+func getallusers(conn *ldap.Conn, UserSearchBaseDNs string, UserSearchFilter string, hello []string) (map[string]string,error){
```
Comment (cviecco) [inline main.go]: function names in CamelCase. Also why are you returning an map of strings? a map of bools is enough.

File: main.go [hunk 16]
```diff
@@ -0,0 +1,463 @@
...
+	final := make(map[string]string)
```
Comment (cviecco) [inline main.go]: to generic name for variable

File: main.go [hunk 17]
```diff
@@ -0,0 +1,463 @@
...
+		return nil,err
+	}
+	fmt.Println(len(result.Entries))
+	if len(result.Entries)==0{
+		log.Println("No records found")
+	}
+	//fmt.Printf("%+v",result.Entries)
+	//fmt.Println(result.Entries[0].GetAttributeValue("uid"))
+	for _,entry := range result.Entries{
+		//if entry.GetAttributeValue("uid")
```
Comment (cviecco) [inline main.go]: remove dead code

File: main.go [hunk 18]
```diff
@@ -0,0 +1,463 @@
...
+			result=append(result,entry)
+		}
+
+	}
+	return result,nil
+}
+
+
+//To build a user base DN using uid only for CPE LDAP.
+func userDN(usersdn string,username string)(string){
```
Comment (cviecco) [inline main.go]: name of function does not match is functionality

File: main.go [hunk 19]
```diff
@@ -0,0 +1,463 @@
...
+	//uid := username
+	result := "uid="+ username +","+usersdn
+
+	return string(result)
+
+}
+
+//To build a GroupDN for a particular group in cpe ldap
+func GroupDN(groupdn string,groupname string)(string){
```
Comment (cviecco) [inline main.go]: same as above, name of function does not map to the functionality

File: main.go [hunk 20]
```diff
@@ -0,0 +1,513 @@
...
+type AppConfigFile struct {
+	Base       baseConfig         `yaml:"base"`
+	SourceLDAP UserInfoLDAPSource `yaml:"source_config"`
+	TargetLDAP UserInfoLDAPSource `yaml:"target_config"`
+}
+
+type RuntimeState struct {
+	Config       AppConfigFile
+	ADldap       *ldap.Conn
+	CPEldap      *ldap.Conn
```
Comment (JohnTheodore) [inline main.go]: Remove all the 'CPE' references. The variable and function names should be suitable for anyone using this codebase. also generally we want scrub localized pronouns in our environment from the codebase on github.

**Comment by JohnTheodore** [comment]
Don't forget to rebase and squash this all into one commit!

**Comment by cviecco** [comment]
will make new PR with extra changes

**Pull request closed without merging**